packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js

```diff
@@ -41,333 +32,320 @@ const DEFAULT_QUERY = {
 	per_page: -1,
 	orderby: 'name',
 	order: 'asc',
-	_fields: 'id,name,parent',
+	context: 'view',
 };
 
 const MIN_TERMS_COUNT_FOR_FILTER = 8;
 
-class HierarchicalTermSelector extends Component {
-	constructor() {
-		super( ...arguments );
-		this.findTerm = this.findTerm.bind( this );
-		this.onChange = this.onChange.bind( this );
-		this.onChangeFormName = this.onChangeFormName.bind( this );
-		this.onChangeFormParent = this.onChangeFormParent.bind( this );
-		this.onAddTerm = this.onAddTerm.bind( this );
-		this.onToggleForm = this.onToggleForm.bind( this );
-		this.setFilterValue = this.setFilterValue.bind( this );
-		this.sortBySelected = this.sortBySelected.bind( this );
-		this.state = {
-			loading: true,
-			availableTermsTree: [],
-			availableTerms: [],
-			adding: false,
-			formName: '',
-			formParent: '',
-			showForm: false,
-			filterValue: '',
-			filteredTermsTree: [],
+/**
+ * Sort Terms by Selected.
+ *
+ * @param {Object[]} termsTree Array of terms in tree format.
+ * @param {number[]} terms     Selected terms.
+ *
+ * @return {Object[]} Sorted array of terms.
+ */
+function sortBySelected( termsTree, terms ) {
+	const treeHasSelection = ( termTree ) => {
+		if ( terms.indexOf( termTree.id ) !== -1 ) {
+			return true;
+		}
+		if ( undefined === termTree.children ) {
+			return false;
+		}
+		return (
+			termTree.children
+				.map( treeHasSelection )
+				.filter( ( child ) => child ).length > 0
+		);
+	};
+	const termOrChildIsSelected = ( termA, termB ) => {
+		const termASelected = treeHasSelection( termA );
+		const termBSelected = treeHasSelection( termB );
+
+		if ( termASelected === termBSelected ) {
+			return 0;
+		}
+
+		if ( termASelected && ! termBSelected ) {
+			return -1;
+		}
+
+		if ( ! termASelected && termBSelected ) {
+			return 1;
+		}
+
+		return 0;
+	};
+	termsTree.sort( termOrChildIsSelected );
```

Comment:
This function mutates the passed argument. Ideally it should be refactored as a pure function (don't touch the passed argument but return a new one)